### PR TITLE
add the configure logic for detecting whether to use named or unnamed semaphores

### PR DESCRIPTION
**Pull request by edgargabriel**
- restructure the code to use HAVE_SEM_OPEN etc. instead of my own macro

### CODE REVIEW

**Comment by edgargabriel** [comment]
@jsquyres would you mind having a brief look at the configure logic for the sharedfp/sm component? The goal is to use named semaphores (sem_open) when available, unnamed semaphores otherwise (sem_init), and disqualify the component if neither of them is available. This is now the correct pull request

**Comment by lanl-ompi** [comment]
Test FAILed.

File: ompi/mca/sharedfp/sm/configure.m4 [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#                        [action-if-cant-compile])
+# ------------------------------------------------
+AC_DEFUN([MCA_ompi_sharedfp_sm_CONFIG],[
+    AC_CONFIG_FILES([ompi/mca/sharedfp/sm/Makefile])
+
+    sharedfp_sm_happy=no
+    sharedfp_sm_open_happy=no
+    sharedfp_sm_init_happy=no
+    AC_CHECK_HEADER([semaphore.h],
+                    [AC_CHECK_FUNCS([sem_open],[sharedfp_sm_open_happy=yes],[])])
```
Comment (jsquyres) [inline ompi/mca/sharedfp/sm/configure.m4]: Minor quibble: you don't need the final empty [] clause -- you can just leave it out.

File: ompi/mca/sharedfp/sm/configure.m4 [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    if test "$sharedfp_sm_open_happy" = "yes" ; then
+	sharedfp_sm_happy="yes"
+    else
+	AC_CHECK_HEADER([semaphore.h],
+            [AC_CHECK_FUNCS([sem_init],[sharedfp_sm_init_happy=yes],[])])
+	
+	if test "$sharedfp_sm_init_happy" = "yes" ; then
+	    sharedfp_sm_happy="yes"
+	fi
+    fi
```
Comment (jsquyres) [inline ompi/mca/sharedfp/sm/configure.m4]: Just curious: is there a reason you used 3 shell variables instead of just 1?

File: ompi/mca/sharedfp/sm/sharedfp_sm_file_open.c [hunk 3]
```diff
@@ -135,7 +135,6 @@ int mca_sharedfp_sm_file_open (struct ompi_communicator_t *comm,
         return OMPI_ERROR;
     }
 
-        free(sm_filename);
```
Comment (jsquyres) [inline ompi/mca/sharedfp/sm/sharedfp_sm_file_open.c]: Is this `free()` supposed to be part of this pull request?  It doesn't seem related to the semaphore stuff.

**Comment by jsquyres** [comment]
@edgargabriel General question: is there a reason you have 2 different mutexes?  (i.e., one in `struct sm_offset` and one in `struct mca_sharedfp_sm_data` (which contains a `struct sm_offset*`)?  This leads to lots of `#if` statements that just select whether to `sem_FOO()` on the one in sm_offset vs. mca_sharedfp_sm_data.

**Comment by jsquyres** [comment]
...also, the name `struct sm_offset` doesn't obey the prefix rule.  :smiling_imp:

File: ompi/mca/sharedfp/sm/sharedfp_sm_file_open.c (hunk 3 above)
Comment (edgargabriel) [inline ompi/mca/sharedfp/sm/sharedfp_sm_file_open.c]: you caught it, wow :-) No, technically its not, its a separate fix, somebody added it to the file (probably as part of the covertiy cleanup), but it is completely wrong, it actually makes this component seg fault. I can commit it separately if you want.

File: ompi/mca/sharedfp/sm/configure.m4 (hunk 2 above)
Comment (edgargabriel) [inline ompi/mca/sharedfp/sm/configure.m4]: no, just found the initial logic simpler than with just one variable.

**Comment by edgargabriel** [comment]
Looking into the mutexes, honestly don't remember of the top of my head. But can probably be merged to a single  one. And I can change the name of the structure

File: ompi/mca/sharedfp/sm/sharedfp_sm_file_open.c (hunk 3 above)
Comment (jsquyres) [inline ompi/mca/sharedfp/sm/sharedfp_sm_file_open.c]: :smile: It's probably fine to include it on this PR, but it's probably a good idea to go ahead and make it a separate commit (ping me offline if you wanna see the git magic to split this out into a separate commit).  Thanks!

**Comment by edgargabriel** [comment]
ok, so I spent quite some time looking into that, and it looks like the two mutex can not easily be merged. The reason is, that one function (sem_init) expects the semaphore to be in the shared memory region, while the other (sem_open) does not.  However, I renamed the structure sm_offset to mca_sharedfp_sm_offset 

Furthermore, while debugging the two scenarios I realized that my previous attempt to merge the two mutexes actually broke the non-named semaphore code. I fixed it right now ( so both named and unnamed semaphore passed the test suite), I am just wandering whether I should truly commit it to this branch, or whether you would like me to do it separately. the same question as with the additional free statement

**Comment by edgargabriel** [comment]
After thinking about that, I think the cleanest way will be to i) close this pull request ii) commit the two fixes to master that are independent of this pull request and iii) generate a new pull request with the configure logic and the macro changes only. It should not take me too long, but it will probably be monday before I get to it. Does this sound ok to you?
